backend/src/main/java/ai/verta/modeldb/project/FutureProjectDAO.java

```diff
@@ -0,0 +1,149 @@
+import ai.verta.modeldb.common.connections.UAC;
+import ai.verta.modeldb.common.futures.FutureGrpc;
+import ai.verta.modeldb.common.futures.FutureJdbi;
+import ai.verta.modeldb.common.futures.InternalFuture;
+import ai.verta.modeldb.exceptions.PermissionDeniedException;
+import ai.verta.modeldb.experimentRun.subtypes.AttributeHandler;
+import ai.verta.modeldb.experimentRun.subtypes.TagsHandler;
+import ai.verta.uac.Action;
+import ai.verta.uac.IsSelfAllowed;
+import ai.verta.uac.ModelDBActionEnum;
+import ai.verta.uac.ResourceType;
+import ai.verta.uac.Resources;
+import ai.verta.uac.ServiceEnum;
+import java.util.Calendar;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.Executor;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class FutureProjectDAO {
+  private static Logger LOGGER = LogManager.getLogger(FutureProjectDAO.class);
+
+  private final Executor executor;
+  private final FutureJdbi jdbi;
+  private final UAC uac;
+
+  private final AttributeHandler attributeHandler;
+  private final TagsHandler tagsHandler;
+
+  public FutureProjectDAO(Executor executor, FutureJdbi jdbi, UAC uac) {
+    this.executor = executor;
+    this.jdbi = jdbi;
+    this.uac = uac;
+
+    String entityName = "ProjectEntity";
+    attributeHandler = new AttributeHandler(executor, jdbi, entityName);
+    tagsHandler = new TagsHandler(executor, jdbi, entityName);
+  }
+
+  public InternalFuture<Void> deleteAttributes(DeleteExperimentRunAttributes request) {
+    final var projectId = request.getId();
+    final var now = Calendar.getInstance().getTimeInMillis();
+
+    final Optional<List<String>> maybeKeys =
+        request.getDeleteAll() ? Optional.empty() : Optional.of(request.getAttributeKeysList());
+
+    return checkProjectPermission(projectId, ModelDBActionEnum.ModelDBServiceActions.UPDATE)
+        .thenCompose(unused -> attributeHandler.deleteKeyValues(projectId, maybeKeys), executor)
+        .thenCompose(unused -> updateModifiedTimestamp(projectId, now), executor);
+  }
+
+  public InternalFuture<List<KeyValue>> getAttributes(GetAttributes request) {
+    final var projectId = request.getId();
+
+    return checkProjectPermission(projectId, ModelDBActionEnum.ModelDBServiceActions.READ)
+        .thenCompose(unused -> attributeHandler.getKeyValues(projectId), executor);
+  }
+
+  public InternalFuture<Void> logAttributes(LogAttributes request) {
+    final var projectId = request.getId();
+    final var attributes = request.getAttributesList();
+    final var now = Calendar.getInstance().getTimeInMillis();
+
+    return checkProjectPermission(projectId, ModelDBActionEnum.ModelDBServiceActions.UPDATE)
+        .thenCompose(unused -> attributeHandler.logKeyValues(projectId, attributes), executor)
+        .thenCompose(unused -> updateModifiedTimestamp(projectId, now), executor);
+  }
+
+  public InternalFuture<Void> addTags(AddExperimentRunTags request) {
```

Comment:
Bad copy and paste

Reply:
Yes, agree it is resolved in each endpoints individual PR https://github.com/VertaAI/modeldb/pull/2119

Reply:
fixed